python_modules/dagster/dagster/_core/asset_graph_view/bfs.py

```diff
@@ -0,0 +1,221 @@
+    tuples with more information about why certain subsets were excluded.
+
+    Visits parents before children.
+    """
+    initial_subsets = list(asset_graph_view.iterate_asset_subsets(initial_asset_graph_subset))
+
+    # invariant: we never consider an asset partition before considering its ancestors
+    queue = ToposortedPriorityQueue(
+        asset_graph_view, initial_subsets, include_full_execution_set=include_full_execution_set
+    )
+
+    visited_graph_subset = initial_asset_graph_subset
+
+    result: AssetGraphSubset = AssetGraphSubset.empty()
+    failed_reasons: Sequence[Tuple[AssetGraphSubset, str]] = []
+
+    asset_graph = asset_graph_view.asset_graph
+
+    while len(queue) > 0:
+        candidate_subset = queue.dequeue()
+        condition_result = condition_fn(candidate_subset, result)
+
+        subset_that_meets_condition = condition_result.passed_asset_graph_subset
+        failed_reasons.extend(condition_result.excluded_asset_graph_subsets_and_reasons)
+
+        result = result | subset_that_meets_condition
+
+        for matching_entity_subset in asset_graph_view.iterate_asset_subsets(
+            subset_that_meets_condition
+        ):
+            # Add any child subsets that have not yet been visited to the queue
+            for child_key in asset_graph.get(matching_entity_subset.key).child_keys:
+                child_subset = asset_graph_view.compute_child_subset(
+                    child_key, matching_entity_subset
+                )
+                unvisited_child_subset = child_subset.compute_difference(
+                    asset_graph_view.get_entity_subset_from_asset_graph_subset(
+                        visited_graph_subset, child_key
+                    )
+                )
+                if not unvisited_child_subset.is_empty:
+                    queue.enqueue(unvisited_child_subset)
+                    visited_graph_subset = (
+                        visited_graph_subset
+                        | AssetGraphSubset.from_entity_subsets([unvisited_child_subset])
+                    )
+
+    return result, failed_reasons
+
+
+def sort_key_for_asset_key(asset_graph: BaseAssetGraph, asset_key: "AssetKey") -> float:
+    """Returns an integer sort key such that asset partition ranges are sorted in
+    the order in which they should be materialized. For assets without a time
+    window partition dimension, this is always 0.
+    Assets with a time window partition dimension will be sorted from newest to oldest, unless they
+    have a self-dependency, in which case they are sorted from oldest to newest.
+    """
+    partitions_def = asset_graph.get(asset_key).partitions_def
+    time_partitions_def = get_time_partitions_def(partitions_def)
+    if time_partitions_def is None:
+        return 0
+
+    # A sort key such that time window partitions are sorted from oldest start time to newest start time
+    start_timestamp = time_partitions_def.start_ts.timestamp
```

Comment:
I'm a bit confused as to how this sort key works. Is it supposed to sort the partitions for a single asset? so that partition `2024-12-01` materializes before `2024-01-01` for `asset_a` (unless it is self dependent)? If so, taking the start time of the `PartitionsDefinition` seems like it wouldn't work because the start time of the definition will be the same for all partitions of the asset. 

Based on the fn comment and how this key is used in the Queue i would expect that it would take an entity subset and base the key off of the start time for the partition range represented by the subset

Reply:
you are right - this made more sense in sort_key_for_asset_partition where you might care about the order in which partitions for a given asset key are visited, but does not really make sense here. Will update, I think this can probably just sort alphabetically by asset key so that there's a consistent ordering within a given level